## Introduction

In this notebook is about ***methods to manipulating the datas*** inside the data frame. Including finding the certain values / positions ( Indexing ) , selecting areas of columns ( Selecting ) , and replace existing values / positions ( Assigning ) .

First of all, let's import our most important tool into Python - *Pandas*

In [26]:
import pandas as pd
print(pd.__version__)

2.2.3


The   `pd.__version__` use as checking the version of pandas, just to make sure it is updated to the latest version.

Of course, without an input csv, it is handy to create a data frame ourselves. Remember, the   `index_col` is to assign the specific column (you define) as the index column.

In [27]:
reviews = pd.read_csv("winemag-data-130k-v2.csv", index_col=0)
pd.set_option('display.max_rows', 5)

The     `pd.set_option()` allows us to ***configure the display*** with desire.

Inside the `pd.set_option()` , `'display.max_rows'` just as the name shows, is the ***maximum number of rows to display***.

Finally, 5 means the maximum number of rows to display is ***total five rows***. (Not counting from the first nor last five rows, but the total.)

In [28]:
reviews

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia
1,Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,87,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Quinta dos Avidagos 2011 Avidagos Red (Douro),Portuguese Red,Quinta dos Avidagos
...,...,...,...,...,...,...,...,...,...,...,...,...,...
129969,France,"A dry style of Pinot Gris, this is crisp with ...",NaN,90,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Marcel Deiss 2012 Pinot Gris (Alsace),Pinot Gris,Domaine Marcel Deiss
129970,France,"Big, rich and off-dry, this is powered by inte...",Lieu-dit Harth Cuvée Caroline,90,21.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...,Gewürztraminer,Domaine Schoffit


The data frame upon shows perfectly about the `pd.set_option(display.max_rows', 5)`, there are total five rows displayed, included the truncated rows (shows as .........)

If the data frame contains more rows than the set limit, Pandas will truncate the output.

In Python and Pandas, to access specific columns in the data frame, we simply calling: `dataframe.columns_name`

it is just like telling Pandas: *"Hey, please help me to find the name of the colum in this dataframe and display it out."*

For example, in the code below. It says: *"Please help me to find the column named country in this dataframe - reviews."*

In [29]:
reviews.country

0            Italy
1         Portugal
            ...   
129969      France
129970      France
Name: country, Length: 129971, dtype: object

Or, we could achieve the same result by using the Python's dictionary access way: 

In [30]:
reviews['country']

0            Italy
1         Portugal
            ...   
129969      France
129970      France
Name: country, Length: 129971, dtype: object

These are the two ways of selecting a specific Series from the data frame. Both of them are optional but when meeting situations like the column names with reserved characters in them.

E.g. if we had a `country providence` column, attribute selection: &nbsp;`reviews.country providence` wouldn't work.

Therefore, using the indexing operator `[]` will be more advantage: 
        `reviews['country providence']` will be executed

To drill down further into specific values in the column, we only need to use the `[]` once more.

In [31]:
reviews['country'][0]

'Italy'

The code tells Pandas: "Hi there, please help me to find the *value of index 0 of the column country* inside the dataframe."

***Note:***  Python always starts counting from 0

| Index  | Value |
|--------|-------|
|   0    | Italy |

## Indexing

Those methods above for indexing in data frame are nice since the principles are same as the Python and it is easy to pick up and use for novice like myself. 

However, Pandas has its own operators for indexing, which is &nbsp; `loc` &nbsp; and &nbsp; `iloc`. These are the advanced operators we are supposed to using.

In [32]:
# Indexing

### `iloc[ ]`

The &nbsp; `iloc` &nbsp; selects data based on the data's numerical position, also known as the ***index - based selection***.

**Appendix:** The name of &nbsp; `iloc` &nbsp; is : &nbsp; ***Integer - location based indexing.***

In [33]:
reviews.iloc[0]

country                                                    Italy
description    Aromas include tropical fruit, broom, brimston...
                                     ...                        
variety                                              White Blend
winery                                                   Nicosia
Name: 0, Length: 13, dtype: object

Typing `reviews.iloc[0]` will select the first row of the data frame.

But why isn't the output should be something like the full value of the column 0 or something else?

Because `iloc` is &nbsp; ***row - first, column - second***. Which is the opposite of the Python :  &nbsp; ***column - first, row - second***.

This means that &nbsp; `iloc`'s &nbsp; **first priority** is to retireve data from data frame's rows, then columns.

Let's take a look below:

In [34]:
reviews.iloc[:, 0]

0            Italy
1         Portugal
            ...   
129969      France
129970      France
Name: country, Length: 129971, dtype: object

The colon `:` inside the square bracket `[ ]` means ***select all rows***.

In a nutshell, the code &nbsp; `reviews.iloc[:, 0]` &nbsp; means select all rows value of the first column ( 0 )

In [35]:
reviews.iloc[:3, 0]

0       Italy
1    Portugal
2          US
Name: country, dtype: object

In this code, the `:3` means select rows until the `set_number - 1`.

In this example, the code will retrieve the &nbsp;  ***first (0), second (1), and third (2)*** &nbsp; rows of datas from column 0

What happens if it is &nbsp; `number_a : number_b` &nbsp; ?

It will retrieve the rows start from ***number_a until number_b - 1***

As the example below:

In [36]:
reviews.iloc[1:3, 0]

1    Portugal
2          US
Name: country, dtype: object

" 1, 2, .....3 and stop, no more data starting from 3"

We could also pass list inside:

In [37]:
reviews.iloc[[0, 1, 2], 0]

0       Italy
1    Portugal
2          US
Name: country, dtype: object

In [38]:
reviews.iloc[0, [0, 1, 2]]

country                                                    Italy
description    Aromas include tropical fruit, broom, brimston...
designation                                         Vulkà Bianco
Name: 0, dtype: object

**Note:** To find the last n elements of the data, we could use &nbsp; `-` &nbsp; to represent.

-5 means the last 5 elements

In [39]:
reviews.iloc[-5:]

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
129966,Germany,Notes of honeysuckle and cantaloupe sweeten th...,Brauneberger Juffer-Sonnenuhr Spätlese,90,28.0,Mosel,NaN,NaN,Anna Lee C. Iijima,NaN,Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...,Riesling,Dr. H. Thanisch (Erben Müller-Burggraef)
129967,US,Citation is given as much as a decade of bottl...,NaN,90,75.0,Oregon,Oregon,Oregon Other,Paul Gregutt,@paulgwine,Citation 2004 Pinot Noir (Oregon),Pinot Noir,Citation
129968,France,Well-drained gravel soil gives this wine its c...,Kritt,90,30.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Gresser 2013 Kritt Gewurztraminer (Als...,Gewürztraminer,Domaine Gresser
129969,France,"A dry style of Pinot Gris, this is crisp with ...",NaN,90,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Marcel Deiss 2012 Pinot Gris (Alsace),Pinot Gris,Domaine Marcel Deiss
129970,France,"Big, rich and off-dry, this is powered by inte...",Lieu-dit Harth Cuvée Caroline,90,21.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...,Gewürztraminer,Domaine Schoffit


### `loc[ ]`

The second method for attribute selection is the &nbsp; `loc` &nbsp; operator: &nbsp; ***Label - based selection***

It is the data index value, rather than the position

In [40]:
reviews.loc[0, 'country']

'Italy'

`iloc` is much simpler than `loc` because it ignores the dataset's indices. &nbsp; We use `iloc` to treat a dataset like a big lists, the only thing we should consider is the position.
On the other hand, `loc` uses the information in the indices. It is usually easier than `iloc` when you have specific indices.

For example, here's one situation where using `loc` is much more easier:

In [41]:
reviews.loc[:, ['taster_name', 'taster_twitter_handle', 'points']]

,taster_name,taster_twitter_handle,points
0,Kerin O’Keefe,@kerinokeefe,87
1,Roger Voss,@vossroger,87
...,...,...,...
129969,Roger Voss,@vossroger,90
129970,Roger Voss,@vossroger,90


If we need to use `iloc` to produce the same results, it would be like: 

In [42]:
reviews.iloc[:, [8, 9, 3]]

,taster_name,taster_twitter_handle,points
0,Kerin O’Keefe,@kerinokeefe,87
1,Roger Voss,@vossroger,87
...,...,...,...
129969,Roger Voss,@vossroger,90
129970,Roger Voss,@vossroger,90


This takes more ***work*** to learn where exactly is those columns in positions.

**Note:** When using `loc` and `iloc`, there's one thing to keep in mind which made these two methods slightly different.
          

`iloc[0:10]` will select entries `0,....,9`, while `loc[0:10]` will select entries `0,....,10`

The core idea are `loc` can include strings. Therefore, we hope to index the required range of entries conveniently. 

If we want to select a certain ranges of alphabet between ***a ~ m***. &nbsp; It is more convenient to use `df.loc[a:m]` rather than `df.loc[a:n]`. (**n** come after **m**)

It is fun and confusing when the data frame is a simple numerical list:

For example, a data frame contains `0,......,1000`.

`df.iloc[0:1000]` will return 1,000 entries, while `df.loc[0:1000]` will return 1,001 entries.    >_<

To get exactly 1,000 entries using `loc`, you will need to lower one digit down and change into: `df.loc[0:999]`

***Because unlike humans, computers always start counting from 0.***

## Manipulatin the index

We can also manipulate the index as we think it fits.

In [43]:
# Manipulating

In [44]:
reviews

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia
1,Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,87,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Quinta dos Avidagos 2011 Avidagos Red (Douro),Portuguese Red,Quinta dos Avidagos
...,...,...,...,...,...,...,...,...,...,...,...,...,...
129969,France,"A dry style of Pinot Gris, this is crisp with ...",NaN,90,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Marcel Deiss 2012 Pinot Gris (Alsace),Pinot Gris,Domaine Marcel Deiss
129970,France,"Big, rich and off-dry, this is powered by inte...",Lieu-dit Harth Cuvée Caroline,90,21.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...,Gewürztraminer,Domaine Schoffit


The `set_index()` can be used to achieve. Here's what happened when we set the `title` column as the index:

In [45]:
reviews.set_index("title")

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,variety,winery
title,,,,,,,,,,,,
Nicosia 2013 Vulkà Bianco (Etna),Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,White Blend,Nicosia
Quinta dos Avidagos 2011 Avidagos Red (Douro),Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,87,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Portuguese Red,Quinta dos Avidagos
...,...,...,...,...,...,...,...,...,...,...,...,...
Domaine Marcel Deiss 2012 Pinot Gris (Alsace),France,"A dry style of Pinot Gris, this is crisp with ...",NaN,90,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Pinot Gris,Domaine Marcel Deiss
Domaine Schoffit 2012 Lieu-dit Harth Cuvée Caroline Gewurztraminer (Alsace),France,"Big, rich and off-dry, this is powered by inte...",Lieu-dit Harth Cuvée Caroline,90,21.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Gewürztraminer,Domaine Schoffit


This is useful when we found some index is better than the current.

## Selection

Selection with conditional is like a ***filter*** to help us select the needed data sets from data frame.

For example, suppose we are interested specifically in better - than average wines produced in Italy.

We could start by checking if each wine is made in italy or not: 

In [46]:
reviews.country == 'Italy'

0          True
1         False
          ...  
129969    False
129970    False
Name: country, Length: 129971, dtype: bool

This operation would return a Series of `True / False` based on the condition matched to the `country` records.

This result could be used inside of the `loc` to select the relevant data:

In [47]:
reviews.loc[reviews.country == 'Italy']

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia
6,Italy,"Here's a bright, informal red that opens with ...",Belsito,87,16.0,Sicily & Sardinia,Vittoria,NaN,Kerin O’Keefe,@kerinokeefe,Terre di Giurfo 2013 Belsito Frappato (Vittoria),Frappato,Terre di Giurfo
...,...,...,...,...,...,...,...,...,...,...,...,...,...
129961,Italy,"Intense aromas of wild cherry, baking spice, t...",NaN,90,30.0,Sicily & Sardinia,Sicilia,NaN,Kerin O’Keefe,@kerinokeefe,COS 2013 Frappato (Sicilia),Frappato,COS
129962,Italy,"Blackberry, cassis, grilled herb and toasted a...",Sàgana Tenuta San Giacomo,90,40.0,Sicily & Sardinia,Sicilia,NaN,Kerin O’Keefe,@kerinokeefe,Cusumano 2012 Sàgana Tenuta San Giacomo Nero d...,Nero d'Avola,Cusumano


The code tells the interpreter: "Hey, please help me to find the data that are relevant to the country Italy."

We also wanted to know the better wines with at least 90 points rated:

In [48]:
reviews.loc[(reviews.country == 'Italy') & (reviews.points >= 90)]

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
120,Italy,"Slightly backward, particularly given the vint...",Bricco Rocche Prapó,92,70.0,Piedmont,Barolo,NaN,NaN,NaN,Ceretto 2003 Bricco Rocche Prapó (Barolo),Nebbiolo,Ceretto
130,Italy,"At the first it was quite muted and subdued, b...",Bricco Rocche Brunate,91,70.0,Piedmont,Barolo,NaN,NaN,NaN,Ceretto 2003 Bricco Rocche Brunate (Barolo),Nebbiolo,Ceretto
...,...,...,...,...,...,...,...,...,...,...,...,...,...
129961,Italy,"Intense aromas of wild cherry, baking spice, t...",NaN,90,30.0,Sicily & Sardinia,Sicilia,NaN,Kerin O’Keefe,@kerinokeefe,COS 2013 Frappato (Sicilia),Frappato,COS
129962,Italy,"Blackberry, cassis, grilled herb and toasted a...",Sàgana Tenuta San Giacomo,90,40.0,Sicily & Sardinia,Sicilia,NaN,Kerin O’Keefe,@kerinokeefe,Cusumano 2012 Sàgana Tenuta San Giacomo Nero d...,Nero d'Avola,Cusumano


By using the `&` will bring the two conditions together. &nbsp; The `&` means `and` in the boolean

Suppose we'll buy any wine that's made in Italy `or` rated above 90 points.

For this, we use a pipe (`|`). &nbsp; This means `or` in the boolean    

In [49]:
reviews.loc[(reviews.country == 'Italy') | (reviews.points >= 90)]

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia
6,Italy,"Here's a bright, informal red that opens with ...",Belsito,87,16.0,Sicily & Sardinia,Vittoria,NaN,Kerin O’Keefe,@kerinokeefe,Terre di Giurfo 2013 Belsito Frappato (Vittoria),Frappato,Terre di Giurfo
...,...,...,...,...,...,...,...,...,...,...,...,...,...
129969,France,"A dry style of Pinot Gris, this is crisp with ...",NaN,90,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Marcel Deiss 2012 Pinot Gris (Alsace),Pinot Gris,Domaine Marcel Deiss
129970,France,"Big, rich and off-dry, this is powered by inte...",Lieu-dit Harth Cuvée Caroline,90,21.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...,Gewürztraminer,Domaine Schoffit


Pandas comes with a few built-in conditional selectors, two of which will highlight here: &nbsp; `isin` &nbsp; & &nbsp; `notnull`

`isin` lets you select data whose values "is in" a list of values. 

For example, we can use it to select wines only from ***Italy or France***:

In [50]:
reviews.loc[reviews.country.isin(['Italy', 'France'])]

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia
6,Italy,"Here's a bright, informal red that opens with ...",Belsito,87,16.0,Sicily & Sardinia,Vittoria,NaN,Kerin O’Keefe,@kerinokeefe,Terre di Giurfo 2013 Belsito Frappato (Vittoria),Frappato,Terre di Giurfo
...,...,...,...,...,...,...,...,...,...,...,...,...,...
129969,France,"A dry style of Pinot Gris, this is crisp with ...",NaN,90,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Marcel Deiss 2012 Pinot Gris (Alsace),Pinot Gris,Domaine Marcel Deiss
129970,France,"Big, rich and off-dry, this is powered by inte...",Lieu-dit Harth Cuvée Caroline,90,21.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...,Gewürztraminer,Domaine Schoffit


Using the boolean condition would achieve the same results, but more complex. 

In [57]:
reviews.loc[(reviews.country == 'Italy') | (reviews.country == 'France')]

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery,critic,index_backwards
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia,everyone,129971
6,Italy,"Here's a bright, informal red that opens with ...",Belsito,87,16.0,Sicily & Sardinia,Vittoria,NaN,Kerin O’Keefe,@kerinokeefe,Terre di Giurfo 2013 Belsito Frappato (Vittoria),Frappato,Terre di Giurfo,everyone,129965
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129969,France,"A dry style of Pinot Gris, this is crisp with ...",NaN,90,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Marcel Deiss 2012 Pinot Gris (Alsace),Pinot Gris,Domaine Marcel Deiss,everyone,2
129970,France,"Big, rich and off-dry, this is powered by inte...",Lieu-dit Harth Cuvée Caroline,90,21.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...,Gewürztraminer,Domaine Schoffit,everyone,1


That's why using the `isin` would be more easier. Imagine you have more than two countries to list out......

The second is `notnull`. It lets you highlight the values which aren't &nbsp; ***empty or not available &nbsp; (NaN values)***. Vise versa for `isnull`

For example, to filter out wines lacking a price tag in the dataset, here's how we could do:

In [52]:
reviews.loc[reviews.price.notnull()]

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
1,Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,87,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Quinta dos Avidagos 2011 Avidagos Red (Douro),Portuguese Red,Quinta dos Avidagos
2,US,"Tart and snappy, the flavors of lime flesh and...",NaN,87,14.0,Oregon,Willamette Valley,Willamette Valley,Paul Gregutt,@paulgwine,Rainstorm 2013 Pinot Gris (Willamette Valley),Pinot Gris,Rainstorm
...,...,...,...,...,...,...,...,...,...,...,...,...,...
129969,France,"A dry style of Pinot Gris, this is crisp with ...",NaN,90,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Marcel Deiss 2012 Pinot Gris (Alsace),Pinot Gris,Domaine Marcel Deiss
129970,France,"Big, rich and off-dry, this is powered by inte...",Lieu-dit Harth Cuvée Caroline,90,21.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...,Gewürztraminer,Domaine Schoffit


"Please list out those wines that have price tags."

## Assigning Data

Going the other way, assigning the data to a Data Frame is easy.

In [53]:
# Assigning

Just think this as &nbsp; ***adding new Series*** &nbsp; in the the exist Data Frame (adding new columns and values):

In [54]:
reviews['critic'] = 'everyone'
reviews['critic']

0         everyone
1         everyone
            ...   
129969    everyone
129970    everyone
Name: critic, Length: 129971, dtype: object

Let's call out the dataframe to see what's added inside:

In [62]:
reviews

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery,critic,index_backwards
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia,everyone,129971
1,Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,87,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Quinta dos Avidagos 2011 Avidagos Red (Douro),Portuguese Red,Quinta dos Avidagos,everyone,129970
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129969,France,"A dry style of Pinot Gris, this is crisp with ...",NaN,90,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Marcel Deiss 2012 Pinot Gris (Alsace),Pinot Gris,Domaine Marcel Deiss,everyone,2
129970,France,"Big, rich and off-dry, this is powered by inte...",Lieu-dit Harth Cuvée Caroline,90,21.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...,Gewürztraminer,Domaine Schoffit,everyone,1


In [72]:
reviews['index_backwards'] = range(len(reviews), 0, -1)
reviews['index_backwards']

0         129971
1         129970
           ...  
129969         2
129970         1
Name: index_backwards, Length: 129971, dtype: int64

The `range(len(reviews), 0, -1)` could be understand in this syntax: &nbsp; `range(start, stop, step)`

Add the number in each index in this sequence: numbers from total counts of reviews (129971) to 0, but decrement by 1.